src/drivers/distance_sensor/vl53lxx/vl53lxx.cpp

```diff
@@ -0,0 +1,1163 @@
+{
+	unsigned count = buflen / sizeof(struct distance_sensor_s);
+	struct distance_sensor_s *rbuf = reinterpret_cast<struct distance_sensor_s *>(buffer);
+	int ret = 0;
+
+	/* buffer must be large enough */
+	if (count < 1) {
+		return -ENOSPC;
+	}
+
+	/* if automatic measurement is enabled */
+	if (_measure_ticks > 0) {
+
+		/*
+		 * While there is space in the caller's buffer, and reports, copy them.
+		 * Note that we may be pre-empted by the workq thread while we are doing this;
+		 * we are careful to avoid racing with them.
+		 */
+		while (count--) {
+			if (_reports->get(rbuf)) {
+				ret += sizeof(*rbuf);
+				rbuf++;
+			}
+		}
+
+		/* if there was no data, warn the caller */
+		return ret ? ret : -EAGAIN;
+	}
+
+	/* manual measurement - run one conversion */
+	do {
+		_reports->flush();
+
+		/* trigger a measurement */
+		if (OK != measure()) {
+			ret = -EIO;
+			break;
+		}
+
+		while (!_collect_phase);
+
+		/* run the collection phase */
+		if (OK != collect()) {
+			ret = -EIO;
+			break;
+		}
+
+		/* state machine will have generated a report, copy it out */
+		if (_reports->get(rbuf)) {
+			ret = sizeof(*rbuf);
+		}
+
+	} while (0);
+
+	return ret;
+}
+
+
+int
+VL53LXX::readRegister(uint8_t reg_address, uint8_t &value)
+{
+	int ret;
+	uint8_t val = 0;
+
+	ret = transfer(&reg_address, sizeof(reg_address), nullptr, 0);
+
+	if (OK != ret) {
+		DEVICE_LOG("i2c::transfer returned %d", ret);
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	/* read from the sensor */
+	ret = transfer(nullptr, 0, &val, 1);
+
+	if (ret < 0) {
+		DEVICE_LOG("error reading from sensor: %d", ret);
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	ret = OK;
+	value = val;
+
+	return ret;
+
+}
+
+
+int
+VL53LXX::readRegisterMulti(uint8_t reg_address, uint8_t *value, uint8_t length)
+{
+	int ret;
+	uint8_t val[6] = {0, 0, 0, 0, 0, 0};
+	ret = transfer(&reg_address, 1, nullptr, 0);
+
+	if (OK != ret) {
+		DEVICE_LOG("i2c::transfer returned %d", ret);
+		return ret;
+	}
+
+	/* read from the sensor */
+	ret = transfer(nullptr, 0, &val[0], length);
+
+	if (ret < 0) {
+		DEVICE_LOG("error reading from sensor: %d", ret);
+		perf_count(_comms_errors);
+		return ret;
+	}
+
+	memcpy(&value[0], &val[0], length);
+
+	ret = OK;
+
+	return ret;
+
+}
+
+
+int
+VL53LXX::writeRegister(uint8_t reg_address, uint8_t value)
+{
+	int ret;
+	uint8_t cmd[2] = {0, 0};
+
+	cmd[0] = reg_address;
+	cmd[1] = value;
+
+	ret = transfer(&cmd[0], 2, nullptr, 0);
+
+	if (OK != ret) {
+		perf_count(_comms_errors);
+		DEVICE_LOG("i2c::transfer returned %d", ret);
+		return ret;
+	}
+
+	ret = OK;
+
+	return ret;
+
+}
+
+
+int
+VL53LXX::writeRegisterMulti(uint8_t reg_address, uint8_t *value,
+			    uint8_t length) //	bytes are send in order as they are in the array
+{
+	// be careful for uint16_t to send first higher byte
+	int ret;
+	uint8_t cmd[6] = {0, 0, 0, 0, 0, 0};
+
+	cmd[0] = reg_address;
+
+	memcpy(&cmd[1], &value[0], length);
```

Comment:
This needs a length check, or an API documentation for `writeRegisterMulti` about the maximum length (`length` must not be more than 5).